### ISSUE TITLE
[BUG] Training Step Time Increases after Function Call

### ISSUE THREAD

**Issue by kimbochen** [ml-explore/mlx#999]
**Describe the bug**

I wrote a model and a training step function.  
First function call takes 8 seconds, and the second takes 1 minute. Not sure why that happens.

**To Reproduce**

Code is in [this repo](https://github.com/kimbochen/mini-llama-mlx/)
To reproduce the issue, execute
```bash
pip install -r requirements.txt
python reproduce_bug.py
```
![mlx-profiling](https://github.com/ml-explore/mlx/assets/40498137/5696e7e4-1628-4253-86f7-691d904040cf)

**Expected behavior**
I expect the first call to take longer because of compilation, but didn't expect the second one to take that long.

**Desktop (please complete the following information):**
 - OS Version: macOS Sonoma 14.4.1
 - Version:
  ```
  mlx==0.9.1
  mlx-lm==0.7.0
  sentencepiece==0.2.0
  notebook==7.1.2
  ipywidgets==8.1.2
  datasets==2.18.0
  asitop==0.0.24
  ```
- Hardware: M1 MacBook Air / 16GB Memory

**Additional context**
- May be caused by memory swapping, but I don't know how to verify.

**Comment by awni** [comment]
Your code runs fine on an M1 Max with 32 GB.

It requires about 25 GB peak memory which definitely suggests that it is swapping.

Also it's good practice to evaluate the state. The reason the first iteration seems fast is because you only evaluate the forward pass to get the loss (not the full forward/backward).

Here's how I modified your loop slightly to log the memory use and evaluate the `state`:

```python
    for _ in range(100):
        tic = time.perf_counter()
        loss = train_step(input_idxs, label_idxs)
        mx.eval(state)
        print(f"Peak memory {mx.metal.get_peak_memory() / 2**30} (GB)")
        print(f'{loss.item()=:.4f}')
        toc = time.perf_counter()
        print(f'{(toc-tic):.4f} s')
```

**Comment by kimbochen** [comment]
It's working now, thank you!
